### ISSUE TITLE
CloudCachePersistentStorageTests.TestReadChecksumProducesLastWrittenChecksum1 failed in CI:  Pipelines - Run 20210715.30 (azure.com)

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#54869]
https://dev.azure.com/dnceng/public/_build/results?buildId=1240966&view=ms.vss-test-web.build-test-results-tab&runId=36910020&resultId=215221&paneView=debug

According to infra, this has failed 4 times in the last 30 days.

Failure is:

```
Assert.False() Failure
Expected: False
Actual:   True
```

The code for this is here:

https://github.com/dotnet/roslyn/blob/dc3d67460c5632f9ea75df8ed098dc319630ec62/src/VisualStudio/CSharp/Test/PersistentStorage/AbstractPersistentStorageTests.cs#L378-L394

So we're in a state where we're overwriting an existing checksum with a null checksum.  However, very rarely there seems to be a race and reading the checksum back out again returns the original that should have been wiped.

**Comment by CyrusNajmabadi** [comment]
@AArnott It feels highly unlikely to me that this is in teh roslyn code.  At teh end of the day all we end up doing is:

Writing the key + checksum:

```C#
                using var bytes = s_byteArrayPool.GetPooledObject();
                checksum.WriteTo(bytes.Object);

                return await WriteStreamAsync(new CacheItemKey(containerKey.Value, name) { Version = bytes.Object }, stream, cancellationToken).ConfigureAwait(false);
```

Then just writing the key with no checksum

```c#
            if (checksum == null)
            {
                return await WriteStreamAsync(new CacheItemKey(containerKey.Value, name), stream, cancellationToken).ConfigureAwait(false);
            }
```

Then trying to read the key/checksum pair.

```c#
            using var bytes = s_byteArrayPool.GetPooledObject();
            checksum.WriteTo(bytes.Object);

            return await _cacheService.CheckExistsAsync(new CacheItemKey(containerKey.Value, name) { Version = bytes.Object }, cancellationToken).ConfigureAwait(false);
```

If htis fails intermittently, it seems almost certainly like it's in the CloudCache lib.  Thanks!

**Comment by AArnott** [comment]
I wonder if this is somehow related to a bug [we fixed recently](https://dev.azure.com/devdiv/DevDiv/_git/VS.CloudCache/pullrequest/323982) where we allowed retrieval of older values given a checksum that was still on disk. It doesn't sound the same, but it feels like that same genre of bug. What version of the library are you running against?

**Comment by CyrusNajmabadi** [comment]
It's possible.  We're on: 17.0.13-alpha  Shoudl we upgrade to something else?

**Comment by CyrusNajmabadi** [comment]
Dotnet upgrade shows no new builds for that @AArnott 

```
   Top-level Package                                           Requested                     Resolved                      Latest
   > Microsoft.VisualStudio.Cache                              17.0.13-alpha                 17.0.13-alpha                 17.0.13-alpha
```

**Comment by AArnott** [comment]
17.0.13-alpha was the latest till yesterday. The newer version would *not* fix this though, so this isn't an issue that upgrading will fix. Can you file an internal devdiv bug to track? Use the "DevDiv\VS Rich Navigation\Cache Service" area path and assign directly to me.

**Comment by AArnott** [comment]
Looking again, this looks identical to the bug I fixed in the PR linked earlier. The first version with this fix is 17.0.9-alpha. Are you sure that the failures you saw happened with that version or later? Might you have updated to 17.0.9-alpha or later *after* all the failures?

Can you compare your test to the [test added here](https://dev.azure.com/devdiv/DevDiv/_git/VS.CloudCache/pullrequest/323982?path=%2Ftest%2FMicrosoft.VisualStudio.Cache.Tests%2FCacheServiceRuntimeTests.cs&_a=files) to see if you can see any differences?

**Comment by CyrusNajmabadi** [comment]
Ah, this may be because in this branch we're referencing 16.10.40-alpha.  Let me try updating htat to 17.x.  Hopefully that won't cause any dependency issues.  Do you have a 16.x line that contains this fix if we needed that?
